### PR TITLE
Fix duplicate node name in graph for tf.concat

### LINKED ISSUE #31137: tf.concat throws error after another call of tf.concat if values is a single Tensor or a list of length 1
**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): no
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): macOS 10.13.6
- TensorFlow installed from (source or binary): from pip install
- TensorFlow version (use command below): v2.0.0-beta0-16-g1d91213fe7 2.0.0-beta1
- Python version: v3.6.7:6ec5cf24b7, Oct 20 2018, 03:02:14

**Describe the current behavior**

Assume that I have already made a previous call of `tf.concat`. When calling again `tf.concat` with a `values` argument which is either a list of `Tensor` objects of length 1 or a single `Tensor` object, I get the following error:

```Duplicate node name in graph: 'concat'```

It seems to be a naming conflict.

**Describe the expected behavior**

From [tf.concat documentation](https://www.tensorflow.org/versions/r2.0/api_docs/python/tf/concat?hl=en):
> `values`: A list of `Tensor` objects or a single `Tensor`.

`tf.concat` should not throw an error when `values` is a list of length 1 or a single `Tensor` object and no naming conflict should rise.

**Code to reproduce the issue**

```
import tensorflow as tf
from tensorflow.keras import Input


print('Using Tensorflow version {} (git version {})'.format(tf.version.VERSION, tf.version.GIT_VERSION))

i = Input(shape=3)
j = Input(shape=4)
try:
    print(tf.concat([i, j], axis=-1))
except Exception as e:
    print(type(e))
    print(e)
try:
    print(tf.concat([i, j], axis=-1))
except Exception as e:
    print(type(e))
    print(e)
try:
    print(tf.concat([i], axis=-1))
except Exception as e:
    print(type(e))
    print(e)
try:
    print(tf.concat(i, axis=-1))
except Exception as e:
    print(type(e))
    print(e)
```

which outputs:
```
Using Tensorflow version 2.0.0-beta1 (git version v2.0.0-beta0-16-g1d91213fe7)
Tensor("concat:0", shape=(None, 7), dtype=float32)
Tensor("concat_1:0", shape=(None, 7), dtype=float32)
<class 'ValueError'>
Duplicate node name in graph: 'concat'
<class 'ValueError'>
Duplicate node name in graph: 'concat'
```

If I comment both `print(tf.concat([i, j], axis=-1))` lines then `print(tf.concat([i], axis=-1))` does not fail, but `print(tf.concat(i, axis=-1))` do. If I also comment `print(tf.concat([i], axis=-1))` then `print(tf.concat(i, axis=-1))` resolves without error.

**Other info / logs**

Full error log:
```
---------------------------------------------------------------------------
InvalidArgumentError                      Traceback (most recent call last)
~/Documents/beta1/lib/python3.6/site-packages/tensorflow/python/framework/ops.py in _create_c_op(graph, node_def, inputs, control_inputs)
   1550   try:
-> 1551     c_op = c_api.TF_FinishOperation(op_desc)
   1552   except errors.InvalidArgumentError as e:

InvalidArgumentError: Duplicate node name in graph: 'concat'

During handling of the above exception, another exception occurred:

ValueError                                Traceback (most recent call last)
<ipython-input-2-bd15a2b20363> in <module>
----> 1 print(tf.concat(i, axis=-1))

~/Documents/beta1/lib/python3.6/site-packages/tensorflow/python/util/dispatch.py in wrapper(*args, **kwargs)
    178     """Call target, and fall back on dispatchers if there is a TypeError."""
    179     try:
--> 180       return target(*args, **kwargs)
    181     except (TypeError, ValueError):
    182       # Note: convert_to_eager_tensor currently raises a ValueError, not a

~/Documents/beta1/lib/python3.6/site-packages/tensorflow/python/ops/array_ops.py in concat(values, axis, name)
   1282           dtype=dtypes.int32).get_shape().assert_is_compatible_with(
   1283               tensor_shape.scalar())
-> 1284       return identity(values[0], name=scope)
   1285   return gen_array_ops.concat_v2(values=values, axis=axis, name=name)
   1286 

~/Documents/beta1/lib/python3.6/site-packages/tensorflow/python/util/dispatch.py in wrapper(*args, **kwargs)
    178     """Call target, and fall back on dispatchers if there is a TypeError."""
    179     try:
--> 180       return target(*args, **kwargs)
    181     except (TypeError, ValueError):
    182       # Note: convert_to_eager_tensor currently raises a ValueError, not a

~/Documents/beta1/lib/python3.6/site-packages/tensorflow/python/ops/array_ops.py in identity(input, name)
     84       return copied
     85   else:
---> 86     ret = gen_array_ops.identity(input, name=name)
     87     # Propagate handle data for happier shape inference for resource variables.
     88     if hasattr(input, "_handle_data"):

~/Documents/beta1/lib/python3.6/site-packages/tensorflow/python/ops/gen_array_ops.py in identity(input, name)
   4251   # Add nodes to the TensorFlow graph.
   4252   _, _, _op = _op_def_lib._apply_op_helper(
-> 4253         "Identity", input=input, name=name)
   4254   _result = _op.outputs[:]
   4255   _inputs_flat = _op.inputs

~/Documents/beta1/lib/python3.6/site-packages/tensorflow/python/framework/op_def_library.py in _apply_op_helper(self, op_type_name, name, **keywords)
    786         op = g.create_op(op_type_name, inputs, dtypes=None, name=scope,
    787                          input_types=input_types, attrs=attr_protos,
--> 788                          op_def=op_def)
    789       return output_structure, op_def.is_stateful, op
    790 

~/Documents/beta1/lib/python3.6/site-packages/tensorflow/python/framework/func_graph.py in create_op(***failed resolving arguments***)
    463     return super(FuncGraph, self).create_op(
    464         op_type, inputs, dtypes, input_types, name, attrs, op_def,
--> 465         compute_device=compute_device)
    466 
    467   def capture(self, tensor, name=None):

~/Documents/beta1/lib/python3.6/site-packages/tensorflow/python/util/deprecation.py in new_func(*args, **kwargs)
    505                 'in a future version' if date is None else ('after %s' % date),
    506                 instructions)
--> 507       return func(*args, **kwargs)
    508 
    509     doc = _add_deprecated_arg_notice_to_docstring(

~/Documents/beta1/lib/python3.6/site-packages/tensorflow/python/framework/ops.py in create_op(***failed resolving arguments***)
   3294           input_types=input_types,
   3295           original_op=self._default_original_op,
-> 3296           op_def=op_def)
   3297       self._create_op_helper(ret, compute_device=compute_device)
   3298     return ret

~/Documents/beta1/lib/python3.6/site-packages/tensorflow/python/framework/ops.py in __init__(self, node_def, g, inputs, output_types, control_inputs, input_types, original_op, op_def)
   1712           op_def, inputs, node_def.attr)
   1713       self._c_op = _create_c_op(self._graph, node_def, grouped_inputs,
-> 1714                                 control_input_ops)
   1715 
   1716     # Initialize self._outputs.

~/Documents/beta1/lib/python3.6/site-packages/tensorflow/python/framework/ops.py in _create_c_op(graph, node_def, inputs, control_inputs)
   1552   except errors.InvalidArgumentError as e:
   1553     # Convert to ValueError for backwards compatibility.
-> 1554     raise ValueError(str(e))
   1555 
   1556   return c_op

ValueError: Duplicate node name in graph: 'concat'
```

### PR DESCRIPTION
**Pull request by yongtang**
This fix tries to address the issue raised in #31137 where
`Duplicate node name in graph: 'concat'` shown up with the
following:
```
import tensorflow as tf
from tensorflow.keras import Input

print('Using Tensorflow version {} (git version {})'.format(tf.version.VERSION, tf.version.GIT_VERSION))

i = Input(shape=3)
j = Input(shape=4)
try:
    print(tf.concat([i, j], axis=-1))
except Exception as e:
    print(type(e))
    print(e)
try:
    print(tf.concat([i, j], axis=-1))
except Exception as e:
    print(type(e))
    print(e)
try:
    print(tf.concat([i], axis=-1))
except Exception as e:
    print(type(e))
    print(e)
try:
    print(tf.concat(i, axis=-1))
except Exception as e:
    print(type(e))
    print(e)
```

The issue was that `identity` node passed the scope as the name.
This fix fixes the issue.

This fix fixes #31137.

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

### CODE REVIEW

File: tensorflow/python/ops/array_ops.py [hunk 1]
```diff
@@ -1296,20 +1296,20 @@ def concat(values, axis, name="concat"):
...
+    # TODO(mrry): Change to return values?
+    if len(values) == 1:  # Degenerate case of one tensor.
+      # Make a throwaway call to convert_to_tensor to make sure
+      # that axis is of the correct type, and make sure that
+      # the returned tensor is a scalar.
+      # TODO(keveman): Implement a standalone type and shape checker.
       ops.convert_to_tensor(
           axis, name="concat_dim",
           dtype=dtypes.int32).get_shape().assert_has_rank(0)
-      return identity(values[0], name=scope)
```
Comment (alextp) [inline tensorflow/python/ops/array_ops.py]: Wouldn't a simpler fix be just to pass name=name to identity?

Comment (yongtang) [reply tensorflow/python/ops/array_ops.py]: Thanks @alextp, indeed that should fix the issue. The PR has been updated.

**Comment by yongtang** [comment]
Thanks @alextp for the review and suggestion. The PR has been updated. Please take a look.

**Review by alextp — DISMISSED** [review]

**Comment by yongtang** [comment]
@alextp @rthadur one kokoro CI test was failing, because of the name change where 'concat:0' =>`concat/concat:0`. I have updated the PR and fixed the failed test.

Please take a look and sorry for the additional inconvenience.

**Review by alextp — APPROVED** [review]

**Pull request merged**